Merge new Memory Stream implementation

sanjeevm0 opened the pull request:
There are considerable number of changes in this request.  Some reorganization of code is also there, and there were some conflicts, including some when merging with Bruno's latest changes.  If you see any issues, please let me know.  I have several things which I will look at, but wanted to merge, otherwise there is a lot of effort in continuing to have to merge large number of changes.
1. I have still to move the array serialization used for the sort benchmark to custom serialization
2. I will modify StreamBase and its derivatives to implement IDisposable instead of ref counting and use replication instead of assignment (overlap with Jin's current work likely)
3. The refactoring overlaps with Bruno's refactoring of MemStream, may need second look.
4. For some reason, using our threadpool for networking code causes UT hangs on Azure VM.  Using system threadpool works fine - so I changed networking code to use system threadpool.  This is likely due to some remaining bugs in threadpool which need to be fixed.  It is only a two line change to switch back.

zhuweirong commented:
This pull requests brings in many commits that will go into master's commit history. Some of them are probably not in working condition.  Could you please consider squash the commits into one or a few meaningful ones for the commit history?

zhuweirong commented on Prajna.sln:
```diff
@@ -120,10 +120,6 @@ Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "Common", "samples\examples\
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "ServiceEndpoint", "src\ServiceLib\ServiceEndpoint\ServiceEndpoint.fsproj", "{45959765-A1FA-4277-BC68-024CAD648546}"
-EndProject
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "Gateway", "src\ServiceLib\Gateway\Gateway.fsproj", "{C588134A-466C-4453-89AB-7DD7CEC58F06}"
-EndProject
```
Are these two projects removed accidentally?

zhuweirong commented on src/tools/tools/file.fs:
```diff
@@ -228,17 +228,17 @@ module  FileTools =
+            | e -> 
+                message <- e.Message
+                bExist <- true
+            //Logger.LogF(LogLevel.MediumVerbose,  fun _ -> sprintf "WriteBytesToFileConcurrentPCompare: complete create and write file '%s' (exists = %b)" filename bExist)
```
why this line was commented out? It was added intentionally to help track down an issue.

zhuweirong replied:
I guess it's because there're compilation error about line 241 with VS2013?  We didn't get it in other places, since we have switched to VS2015.  It seems the new compiler relaxed the capture requirements of closure.  We probably should make it correct for VS2013 too.

zhuweirong commented on src/tools/tools/file.fs:
```diff
@@ -309,7 +309,7 @@ module  FileTools =
                             else
                                 Threading.Thread.Sleep( 5 )
-            Logger.LogF(LogLevel.MediumVerbose,  fun _ -> sprintf "WriteBytesToFileConcurrentPCompare: done verify file '%s' (verified = %d), detail: %s" filename nVerified message)
+            //Logger.LogF(LogLevel.MediumVerbose,  fun _ -> sprintf "WriteBytesToFileConcurrentPCompare: done verify file '%s' (verified = %d), detail: %s" filename nVerified message)
```
why this line was commented out? It was added intentionally to help track down an issue.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -875,20 +897,25 @@ and [<AllowNullLiteral>]
-                                         ) 
-                        //let bCansend = peerQueue.CanSend 
+//                    if Interlocked.CompareExchange (peerQueue.flowcontrol_lock,1,0) = 0 then
+////                        #tag: disable flowcontrol
```
No need to take the lock any more?

sanjeevm0 replied:
Yes, flow control is disabled here, as network handles all flow control

zhuweirong commented on src/CoreLib/DSetGenerics.fs:
```diff
@@ -626,6 +628,26 @@ type DSet<'U> () =
+        // x.SerializationLimit <- x.SerializationLimit
+        x
+
+    /// <summary>
```
For introducing a new DSet API here, there are a few more places to change
1. Add a corresponding static member in this file
2. Add the corresponding C# API in DSetCSharp.fs

sanjeevm0 replied:
Done

zhuweirong commented on src/CoreLib/GV.fs:
```diff
@@ -1190,7 +1194,7 @@ type internal GVSerialize<'K>( ) =
         let wrapperFunc (ms:MemStream) (O1:Object) =
             let state1 = O1 :?> 'K
-            ms.SerializeFrom( state1 )
+            Strm.SerializeFrom( ms, state1 )
```
Is "Strm" a type name? What does it stands for?  In general, .Net framework guideline does not prefer using abbreviations for naming unless it's very well known (for example IO).   See:  https://msdn.microsoft.com/en-us/library/141e06ef(v=vs.71).aspx

zhuweirong commented on src/CoreLib/deploy.fs:
```diff
@@ -217,10 +219,10 @@ type DeploymentSettings() =
     static member val internal DefaultCacheSerializationLimit = 1 with get, set
     /// ReSerialization For Cache
-    static member val internal DefaultIOMaxQueue = 10 with get, set
+    static member val internal DefaultIOMaxQueue = 100 with get, set
```
Is this a good default for general use or it's optimized towards sort benchmark?

sanjeevm0 replied:
I think it is fine to use this as memory is controlled by the shared stack.  Using higher number may in some cases increase delay on per message basis, but most of our current workload is not sensitive to this.

zhuweirong commented on src/CoreLib/deploy.fs:
```diff
@@ -256,9 +258,27 @@ type DeploymentSettings() =
     static member val MaxSendingQueueLimit = 1024 * 1024 * 50 with get, set
     /// Maximum network stack memory (total across all connections), 0 means unbounded (in bytes)
-    static member val MaxNetworkStackMemory = 1024 * 1024 * 1024
+    static member val MaxNetworkStackMemory = 1024 * 1024 * 1024 * 4 with get, set
```
Is this a good default for general use or it's optimized towards sort benchmark?

sanjeevm0 replied:
4GB is fine as there is additional control of using certain percentage of available RAM.

zhuweirong commented on src/tools/tools/configuration.fs:
```diff
@@ -244,7 +244,7 @@ module internal ConfigurationUtils =
 
     // Unpack asm binding information
-    let UnpackAsmBinding (ms : MemStream) =
+    let UnpackAsmBinding (ms : StreamBase<byte>) =
```
How about the corresponding PackAsmBinding function, does it also need to change to use StreamBase<byte> instead of MemStream?

jinlmsft replied:
Perhaps PackAsmBinding  and UnpackAsmBinding can consider to use (ms:Stream)?

sanjeevm0 replied:
Changed both to Stream

BrunoBozza commented on samples/DistributedSort/DistributedSort.fs:
```diff
@@ -207,27 +213,34 @@ let main orgargs =
         //let verifyCount = dkvGen |> DSet.fold (fun cnt x -> cnt + 1) (fun a b -> a + b) 0
+        let foldFunc = (fun cnt x ->
+            cnt + 1
+        )
```
It seems like foldFunc is dead. If you'd like to keep it around, you could comment it out and add a comment explaining why it could be useful.

sanjeevm0 replied:
done

BrunoBozza commented on samples/SortBenchmark/DiskHelper.fs:
```diff
@@ -41,13 +41,12 @@ open Prajna.Tools
-    member val drives = DriveInfo.GetDrives()
-    member val diskProfile = DriveInfo.GetDrives() |> Array.filter(fun di -> di.DriveType = DriveType.Fixed)  |> Array.filter(fun di -> di.TotalSize > 100000000000L) |> Array.map(fun di -> Path.Combine(di.Name,"sortbenchmark").ToLower()+"\\")
+    //member val drives = DriveInfo.GetDrives()
+    member val diskProfile = null with get,set
```
Public properties should be use PascalCasing.
Also, what's the type of diskProfile? I was thinking StorageProfile, but that doesn't allow the null literal. Maybe add a type annotation?
Does it need to be initialized with null? In internal F# code, I would make this an option, but since it's a public type, maybe we could use a "NullProfile" object? This would avoid null checking everywhere else.

hongzhili replied:
diskProfile is an array. 
It has to be initialized at remote container in order to get hard drive configuration of the remote machine (not client machine). And yes, a type annotation or comment is needed here.

BrunoBozza replied:
Can it be initialized with a Lazy<'a[]>? If this is meant to be accessed only at the remote machine, a Lazy has another advantage: it let's you throw an informative exception in case one tries to access it at the client. Could be as simple as 'failwith "Shouldn't be accessing this at the client"', which is way better than a NullReferenceException, and making everyone check for null everywhere.
To be clear: I only propose implementing this before the pull if it is trivial and doesn't add any delay to the commit. Otherwise, we should pull this as is and change (or not) later. My general take on code reviews is that, as long as the code is useful and not harmful outside its context (and this clearly qualifies), commited code is always better than code sitting in your repo.

hongzhili replied:
Thanks Bruno for the suggestions. I agree. Since SortBenchmark is an individual sample project. Let's just merge this pull request if all other issues are addressed and I can integrate the comments after it is merged. Actually, it's not easy for me to make any changes right now since it's in Sanjeev's branch.

BrunoBozza commented on samples/SortBenchmark/DiskHelper.fs:
```diff
@@ -291,3 +299,85 @@ type DiskHelper(records:int64) =
+
+
+
+[<AllowNullLiteral>]
```
General comment: I am against "null" whenever possible. In F#, there's option. For public interface code, if really there's no way to construct objects that are "complete", we can use the null object pattern: https://en.wikipedia.org/wiki/Null_Object_pattern. Unfortunately, C# doesn't have a "[DisallowNullLiteral]" attribute, so ultimately one is always exposed. But generally avoiding null pointers at runtime will go a long way towards avoiding NullReferenceException. That's one of the most common kinds of exceptions, and it's a whole class of bugs that can largely mitigated.

hongzhili replied:
I agree. Actually, this type may be removed later since we changed the implementation.

BrunoBozza commented on samples/SortBenchmark/DiskHelper.fs:
```diff
@@ -291,3 +299,85 @@ type DiskHelper(records:int64) =
+
+
+[<AllowNullLiteral>]
+type RollingFileMgr(records:int64) = 
```
Public properties and fields should be PascalCased.

BrunoBozza commented on samples/SortBenchmark/DiskHelper.fs:
```diff
@@ -291,3 +299,85 @@ type DiskHelper(records:int64) =
+
+
+[<AllowNullLiteral>]
+type SingleThreadExec1() =
```
There's almost never a reason for a type with an empty constructor to have an [<AllowNullLiteral>]. To delay initialization, we can always just use a Lazy<'T>, or just create the object if initialization is not an issue.

BrunoBozza commented on samples/SortBenchmark/SortBenchmark.fs:
```diff
@@ -43,40 +43,23 @@ open Prajna.Tools
-    -num        Number of remote instances running \n\
-    -sort       Executing Sort (1: 1-pass sort, 2-2 pass sort) \n\
-    -nump N     Number of partitions \n\
+    -nump       Number of partitions in the second stage repartition\n\
```
We should start using a proper command line parsing library. Ken was working in VMHub with the same one we used in TLC/AzureML Algos. It's pretty good. I'm sure there are others. This should not block this pull, of course, as there's tons of code that's like this and we don't need to change it now. But going forward, we should move towards something like that.

BrunoBozza commented on samples/SortBenchmark/SortBenchmark.fs:
```diff
@@ -97,12 +80,19 @@ module Interop =
+        let pinedbuffer = bufferHandle.AddrOfPinnedObject()        
+        stdqsort(pinedbuffer, buffer.Length / dim, dim)
+        bufferHandle.Free()
+
```
Use try/finally to free the handle.

BrunoBozza commented on samples/SortBenchmark/SortBenchmark.fs:
```diff
@@ -141,148 +131,34 @@ module Interop =
-        let fileReadyToWrite = dataBufSize.ToArray() 
-                                |> Seq.filter (fun kv -> !(kv.Value) > x.WriteBlockSize  ) 
-                                |> Seq.map (fun kv -> kv.Key)
+type RepartitionStage = StageOne=1 | StageTwo=2
```
Does this need to be public?

BrunoBozza commented on samples/SortBenchmark/SortBenchmark.fs:
```diff
@@ -390,98 +266,310 @@ type RemoteFunc( filePartNum:int, records:int64, _dim:int , partNumS1:int, partN
+            let byt = ref Unchecked.defaultof<_>
+            let bnewbuf = ref false
+            if not (RemoteFunc.sharedMem.TryDequeue(byt)) then
+                if (Interlocked.Increment(RemoteFunc.sharedMemSize) < 300) then
```
General comment: There's no need for the extra indirection with "ref". So long as you don't capture the variables (byt and bnewbuf) in closures, you can use just "let mutable" and "&byt" when passing to TryDeque. Probably doesn't make much difference in most code, but just fyi.

BrunoBozza commented on samples/SortBenchmark/SortBenchmark.fs:
```diff
@@ -390,98 +266,310 @@ type RemoteFunc( filePartNum:int, records:int64, _dim:int , partNumS1:int, partN
+                        let remainpos = ref 0
+                        let remainLen = ref 0
+
+                        while !bHasBuf do
```
Seems you really care about perf. I think you'll see some benefit from "let mutable", not to mention less noisy syntax.

BrunoBozza commented on samples/SortBenchmark/SortBenchmark.fs:
```diff
@@ -390,98 +266,310 @@ type RemoteFunc( filePartNum:int, records:int64, _dim:int , partNumS1:int, partN
+
+                                if !bRemainRecord then
+                                    if (!remainLen - !remainpos >=2) then
+                                        let index = (((int) (!remainBuf).[!remainpos]) <<< 8) + ((int) (!remainBuf).[!remainpos + 1])
```
No need to parenthesize the "int" in a typecast, since it's a regular function.

BrunoBozza commented on samples/SortBenchmark/SortBenchmark.fs:
```diff
@@ -552,36 +631,103 @@ type RemoteFunc( filePartNum:int, records:int64, _dim:int , partNumS1:int, partN
                         let partionBoundary =
-                            if stage = 1 then
+                            if stage = RepartitionStage.StageOne then
                                 stageOnePartionBoundary
```
If the stage is never sent over the wire, there's an advantage to _not_ making it an enum (by ommiting the "= 1", "= 2" in the stage definition). The advantage is that the "match" keyword will check that the alternatives are exhaustive at compile time. As this is, if a "StageThree" is ever introduced, we'll have to go hunting for where to update. If we use an F# union and "match", the compiler does that for us.

BrunoBozza commented on src/CoreLib/DSeq.fs:
```diff
@@ -1011,10 +1041,14 @@ and [<AllowNullLiteral>]
-                let msCombine = ms.InsertBefore( msPrefix )
+                //let msCombine = ms.InsertBefore( msPrefix )
+                ms.InsertBefore(msPrefix) |> ignore
+                ms.DecRef()
```
Can we wrap the GetNew()/DecRef() pair into a "using" type function, so the DecRef() is guaranteed. Something like:

`````` F#
let withNew (f: stream -> 'a) ms = 
    let x = ms.GetNew() 
    let ret = f x
    ms.DecRef()
    ret```
``````

BrunoBozza replied:
Same for AddRef.

sanjeevm0 replied:
Yes, Jin is also taking a look at this.  This will eventually happen.

BrunoBozza commented on src/CoreLib/network.fs:
```diff
@@ -170,55 +170,59 @@ type NetworkPerformance() =
+                x.ms.Length - x.startPos
+
+    /// release the underlying memstream
+    member x.Release() =
```
Why not standard IDisposeable with a call to SupressFinalizer in the Dispose() implementation?

sanjeevm0 replied:
I have added a IDisposable, however I have left Release in also.  It is a bit hard to wrap NetworkCommand in a using statement as it is in a queue, and when processing is done, it needs to be released.

BrunoBozza commented on src/CoreLib/network.fs:
```diff
@@ -1001,29 +1099,183 @@ type [<AllowNullLiteral>] NetworkCommandQueue() as x =
-        let rcvdSerialBuf = BitConverter.GetBytes(Operators.uint16(x.RcvdCommandSerial))
-        Buffer.BlockCopy(rcvdSerialBuf, 0, cmd.buffer, 6, 2)
+        let pos = cmd.ms.Position
+        cmd.ms.Seek(6L, SeekOrigin.Begin) |> ignore
```
Why 6? Maybe use symbolic constants?

BrunoBozza commented:
Phew! I'm done. I'll update my generic serialization to use this new code once it goes in. In general, there should be no reason for anyone to write a "custom" encoder/decoder for arrays/matrices unless they know something special about their data (e.g.: image, sound, etc). We could enable general gzip/lzw encoding/decoding as an option, which would go a long way for text. If we also add something lossless for System.Drawing.Bitmap and whatever the .Net PCM waveform object is, we should be in good shape for many workloads with quite good performance.

sanjeevm0 committed:
1. Merge new MemoryStream implementation
Consists of several new types in bufferliststream.fs (e.g., ref counted buffer, and base class StreamBase)
Uses SharedMemoryPool for memory which is reused.
Ref counting allows for stream copy without memory copy.
Buffers are refcounted so multiple streams can share differing portions of same buffer.
Buffers are self-releasing so stream can consist of mixed buffers and reuse other buffers without copy.
2. Also allows us to specify particular IP address to use in case multiple NICs are present on machine.
This is useful in certain scenarios such where one NIC may be connected to fast network connection

Still TODO
1.Remove the array serialization used for the sort benchmark to custom serialization as Bruno's new code
should do array serialization fast.
2.Jin and I will modify StreamBase and its derivatives to implement IDisposable instead of ref counting
and use replication instead of assignment
3.The refactoring overlaps with Bruno's refactoring of MemStream, may need second look.
4.For some reason, using our threadpool for networking code causes UT hangs on Azure VM.
Using system threadpool works fine - so I changed networking code to use system threadpool for now.
This is likely due to some remaining bugs in threadpool which need to be fixed.
It is only a two line change to switch back to using our own threadpool.
